File: Referential.cpp
#include "Referential.h"
#include <cmath>
//////////////////////////////////////////////////////////////////////////////
Referential::Referential()
{
    _rotMat[0]=1.; _rotMat[1]=0.;  _rotMat[2]=0.;
    _rotMat[3]=0.; _rotMat[4]=1.;  _rotMat[5]=0.;
    _rotMat[6]=0.; _rotMat[7]=0.;  _rotMat[8]=1.;

    _x=0.; _y=0.; _z=0.;
}
//////////////////////////////////////////////////////////////////////////////
Referential::~Referential()
{ }
//////////////////////////////////////////////////////////////////////////////
void Referential::setYawPitchRoll(double Yaw,double Pitch,double Roll)
{ // from public domain www.developpez.com
    double A=cos(Roll);
    double B=sin(Roll);
    double C=cos(Pitch);
    double D=sin(Pitch);
    double E=cos(Yaw);
    double F=sin(Yaw);

    _rotMat[0]=C*E;
    _rotMat[1]=-C*F;
    _rotMat[2]=-D;

    _rotMat[3]=-B*D*E + A*F;
    _rotMat[4]= B*D*F + A*E;
    _rotMat[5]=-B*C;

    _rotMat[6]= A*D*E + B*F;
    _rotMat[7]=-A*D*F + B*E;
    _rotMat[8]= A*C;
}
//////////////////////////////////////////////////////////////////////////////
void Referential::transform(double& x,double& y,double& z) const
{
    double xt=x+_x;
    double yt=y+_y;
    double zt=z+_z;

    double nx=_rotMat[0]*xt+_rotMat[1]*yt+_rotMat[2]*zt;
    double ny=_rotMat[3]*xt+_rotMat[4]*yt+_rotMat[5]*zt;
    double nz=_rotMat[6]*xt+_rotMat[7]*yt+_rotMat[8]*zt;

    x=nx; y=ny; z=nz;
}
//////////////////////////////////////////////////////////////////////////////
void Referential::unTransform(double& x,double& y,double& z) const
{
    double nx=_rotMat[0]*x+_rotMat[3]*y+_rotMat[6]*z;
    double ny=_rotMat[1]*x+_rotMat[4]*y+_rotMat[7]*z;
    double nz=_rotMat[2]*x+_rotMat[5]*y+_rotMat[8]*z;

    x=nx-_x; y=ny-_y; z=nz-_z;
}
//////////////////////////////////////////////////////////////////////////////
void Referential::setTranslation(double x,double y, double z)
{
    _x=x; _y=y;_z=z;
}
//////////////////////////////////////////////////////////////////////////////

File: python/Referential.py
import math

class Referential:
    def __init__(self):
        self._rotMat = [1., 0., 0., 0., 1., 0., 0., 0., 1.]
        self._x = 0.
        self._y = 0.
        self._z = 0.

    def setYawPitchRoll(self, Yaw, Pitch, Roll):
        A = math.cos(Roll)
        B = math.sin(Roll)
        C = math.cos(Pitch)
        D = math.sin(Pitch)
        E = math.cos(Yaw)
        F = math.sin(Yaw)

        self._rotMat[0] = C * E
        self._rotMat[1] = -C * F
        self._rotMat[2] = -D

        self._rotMat[3] = -B * D * E + A * F
        self._rotMat[4] = B * D * F + A * E
        self._rotMat[5] = -B * C

        self._rotMat[6] = A * D * E + B * F
        self._rotMat[7] = -A * D * F + B * E
        self._rotMat[8] = A * C

    def transform(self, x, y, z):
        xt = x + self._x
        yt = y + self._y
        zt = z + self._z

        nx = self._rotMat[0] * xt + self._rotMat[1] * yt + self._rotMat[2] * zt
        ny = self._rotMat[3] * xt + self._rotMat[4] * yt + self._rotMat[5] * zt
        nz = self._rotMat[6] * xt + self._rotMat[7] * yt + self._rotMat[8] * zt

        return nx, ny, nz

    def unTransform(self, x, y, z):
        nx = self._rotMat[0] * x + self._rotMat[3] * y + self._rotMat[6] * z
        ny = self._rotMat[1] * x + self._rotMat[4] * y + self._rotMat[7] * z
        nz = self._rotMat[2] * x + self._rotMat[5] * y + self._rotMat[8] * z

        return nx - self._x, ny - self._y, nz - self._z

    def setTranslation(self, x, y, z):
        self._x = x
        self._y = y
        self._z = z